nemoguardrails/actions/hallucination/hallucination.py

```diff
@@ -53,24 +56,46 @@ async def check_hallucination(
     if bot_response and last_bot_prompt_string:
         num_responses = HALLUCINATION_NUM_EXTRA_RESPONSES
         # Use beam search for the LLM call, to get several completions with only one call.
-        # At the current moment, only OpenAI LLM engines are supported for computing the additional completions.
-        if type(llm) != OpenAI:
+        # At the current moment, only OpenAI LLM and OpenAI chat model engines, and their customized sub engines are supported for computing the additional completions.
+        if not isinstance(llm, (OpenAI, ChatOpenAI)):
             log.warning(
-                f"Hallucination rail can only be used with OpenAI LLM engines."
+                f"Hallucination rail can only be used with OpenAI LLM and chat model engines, and their customized sub engines"
                 f"Current LLM engine is {type(llm).__name__}."
             )
             return False
 
+        # Customized sub engine can only be either OpenAI LLM or OpenAI chat model
+        if isinstance(llm, OpenAI) and isinstance(llm, ChatOpenAI):
+            log.warning(
+                f"LLM engine must be either OpenAI LLM or chat model."
+                f"Current LLM engine is a merge of two types."
+            )
+            return False
+
         # Use the "generate" call from langchain to get all completions in the same response.
         last_bot_prompt = PromptTemplate(template="{text}", input_variables=["text"])
+
+        # Create message prompt for chat model
+        if isinstance(llm, ChatOpenAI):
+            human_message_prompt = HumanMessagePromptTemplate(prompt=last_bot_prompt)
+            last_bot_prompt = ChatPromptTemplate.from_messages([human_message_prompt])
+
         chain = LLMChain(prompt=last_bot_prompt, llm=llm)
 
         # Generate multiple responses with temperature 1.
-        with llm_params(llm, temperature=1, n=num_responses, best_of=num_responses):
-            extra_llm_response = await chain.agenerate(
-                [{"text": last_bot_prompt_string}],
-                run_manager=logging_callback_manager_for_chain,
-            )
+        if isinstance(llm, OpenAI):
+            with llm_params(llm, temperature=1, n=num_responses, best_of=num_responses):
+                extra_llm_response = await chain.agenerate(
+                    [{"text": last_bot_prompt_string}],
+                    run_manager=logging_callback_manager_for_chain,
+                )
+        else:
+            # Chat model has no `best_of` argument
+            with llm_params(llm, temperature=1, n=num_responses):
+                extra_llm_response = await chain.agenerate(
+                    [{"text": last_bot_prompt_string}],
+                    run_manager=logging_callback_manager_for_chain,
+                )
```

Comment:
I think here you can use directly the value from `context.get("_last_bot_prompt")` and pass it to `llm_call` (https://github.com/NVIDIA/NeMo-Guardrails/blob/main/nemoguardrails/actions/llm/utils.py#L27) which will deal correctly with chat vs. completion.

Reply:
@trebedea It looks like L86-98 generated a batch of results because of the specified `num_responses`, they were popped out one by one in L105-112. `llm_call` seems to be just for a single result, so it has to be called `num_responses` times in a loop. Is this correct? If yes, I can handle it.

Reply:
The cleanest way to do this is to make `llm_call` return all generations / responses when called with `n>1` (for example, for `OpenAI `and `ChatOpenAI`). I've checked and we have access to all of them, but only return the first one in the current implementation. 

It would be only a couple of lines of changes to make this possible - return a string if only one completion in `result.generations[0]` , otherwise return all completions in a list:
`async def llm_call(llm: BaseLanguageModel, prompt: Union[str, List[dict]]) -> Union[str, List[str]]:`

Do you want to take this?

Reply:
I submitted a new commit to resolve the three changes.